File: src/main/java/jPetrovich/Petrovich.java
package jPetrovich;

/**
 * @author DMITRY KNYAZEV
 * @since 31.05.2014
 */

import jPetrovich.util.CommonUtils;
import jPetrovich.util.ECase;
import jPetrovich.util.EGender;
import jPetrovich.util.rules.RulesLoader;
import jPetrovich.util.rules.data.Rule;
import jPetrovich.util.rules.data.RuleSet;
import jPetrovich.util.rules.data.Rules;
import org.apache.commons.lang3.StringUtils;

import java.io.FileNotFoundException;
import java.util.*;

public class Petrovich {

	private final Rules _rules;

	/**
	 * Use specified rules
	 *
	 * @param rules you rules.
	 * @throws NullPointerException if rules is {@code null}
	 */
	public Petrovich(Rules rules) {
		if (rules == null) throw new NullPointerException("Rules must not be null");
		this._rules = rules;
	}

	/**
	 * Load rules from file path
	 *
	 * @param path file path
	 * @throws FileNotFoundException if file path not exist
	 */
	public Petrovich(String path) throws FileNotFoundException {
		this(RulesLoader.loadRules(path));

	}

	/**
	 * Load rules from default file path.
	 * Default path is "current_dir\rules.yml"
	 *
	 * @throws FileNotFoundException if file path not exist
	 */
	public Petrovich() throws FileNotFoundException {
		this(RulesLoader.loadRules());
	}

	/**
	 * Convert first name from nominative case to specified case
	 *
	 * @param firstName  first name at nominative case
	 * @param genderCase gender
	 * @param nameCase   case
	 * @return converted first name
	 */
	public String firstName(String firstName, EGender genderCase, ECase nameCase) {
		return convertTo(firstName, genderCase, nameCase, getRules().firstname);
	}

	/**
	 * Convert last name from nominative case to specified case
	 *
	 * @param lastName   last name at nominative case
	 * @param genderCase gender
	 * @param nameCase   case
	 * @return converted last name
	 */
	public String lastName(String lastName, EGender genderCase, ECase nameCase) {
		return convertTo(lastName, genderCase, nameCase, getRules().lastname);
	}

	/**
	 * Convert middle name from nominative case to specified case
	 *
	 * @param middleName middle name at nominative case
	 * @param genderCase gender
	 * @param nameCase   case
	 * @return converted middle name
	 */
	public String middleName(String middleName, EGender genderCase, ECase nameCase) {
		return convertTo(middleName, genderCase, nameCase, getRules().middlename);
	}

	/**
	 * Convert middle name from nominative case to specified case with auto detect gender
	 *
	 * @param middleName middle name at nominative case
	 * @param nameCase   case
	 * @return converted middle name
	 */
	public String middleName(String middleName, ECase nameCase) {
		EGender genderCase = CommonUtils.detectGender(middleName);
		return middleName(middleName, genderCase, nameCase);
	}

	private String convertTo(String sourceName, EGender genderCase, ECase nameCase, RuleSet ruleSet) {
		String[] splittedName = sourceName.split("-");
		List<String> result = new ArrayList<>(splittedName.length);
		for (int index = 0; index < splittedName.length; index++) {
			final boolean firstWord = index == 0 && splittedName.length > 1;
			Map<String, Boolean> features = new HashMap<String, Boolean>() {{
				put("first_word", firstWord);
			}};
			String modifiedWord = findAndApply(splittedName[index], genderCase, nameCase, ruleSet, features);
			result.add(modifiedWord);
		}
		return StringUtils.join(result, "-");
	}

	private String findAndApply(String name, EGender genderCase, ECase nameCase, RuleSet ruleSet, Map<String, Boolean> features) {
		Rule rule = findRule(name, genderCase, ruleSet, features);

		if (rule == null)
			return name;

		return apply(name, nameCase, rule);
	}

	private Rule findRule(String name, EGender genderCase, RuleSet ruleSet, Map<String, Boolean> features) {
		Set<String> tags = features.keySet();

		Rule rule;
		if (ruleSet.exceptions != null) {
			rule = findRule(name, genderCase, ruleSet.exceptions, true, tags);
			if (rule != null)
				return rule;
		}
		return findRule(name, genderCase, ruleSet.suffixes, false, tags);
	}

	private Rule findRule(String name, EGender genderCase, Rule[] rules, Boolean matchWholeWord, Set<String> tags) {
		for (Rule rule : rules) {
			if (matchRule(name, genderCase, rule, matchWholeWord, tags))
				return rule;
		}
		return null;
	}

	private boolean matchRule(String name, EGender genderCase, Rule rule, Boolean matchWholeWord, Set<String> tags) {
		//todo simplify
		List<String> ex = new ArrayList<>();
		if (rule.tags != null) ex.addAll(Arrays.asList(rule.tags));
		ex.removeAll(tags);

		if (!ex.isEmpty())
			return false;

		EGender genderRule = EGender.getGender(rule.gender);
		if (EGender.getGender(rule.gender) != null &&
				((genderRule == EGender.male && genderCase == EGender.female) ||
						(genderRule == EGender.female && genderCase != EGender.female))) {
			return false;
		}

		name = name.toLowerCase();
		for (String chars : rule.test) {
			String test = matchWholeWord ? name : name.substring(Math.max(0, name.length() - chars.length()));
			if (test.equals(chars))
				return true;
		}
		return false;
	}

	private String apply(String name, ECase nameCase, Rule rule) {
		for (char str : FindCaseModifier(nameCase, rule).toCharArray()) {
			switch (str) {
				case '.':
					break;
				case '-':
					name = name.substring(0, name.length() - 1);
					break;
				default:
					name += str;
					break;
			}
		}
		return name;
	}

	private String FindCaseModifier(ECase nameCase, Rule rule) {
		switch (nameCase) {
			case nominative:
				return "";
			case genitive:
				return rule.mods[0];
			case dative:
				return rule.mods[1];
			case accusative:
				return rule.mods[2];
			case instrumental:
				return rule.mods[3];
			case prepositional:
				return rule.mods[4];
			default:
				throw new IllegalArgumentException(String.format("Unknown grammatical case: %s", nameCase));
		}
	}

	//getters/setters
	public Rules getRules() {
		return _rules;
	}

}

File: src/main/kotlin/kt/petrovich/Petrovich.kt
package kt.petrovich

import kt.petrovich.rules.Rule
import kt.petrovich.rules.RuleGroup
import kt.petrovich.rules.Rules
import java.io.FileNotFoundException
import java.util.ArrayList

/**
 * @author Dmitrii Kniazev
 * @since 31.05.2014
 */
class Petrovich(private val rules: Rules) {

    /**
     * Load rules from file path
     *
     * @param path file path
     * @throws FileNotFoundException if file path not exist
     */
    @Throws(FileNotFoundException::class)
    public constructor(path: String) : this(Rules.loadRules(path))

    /**
     * Load rules from default file from resources.
     */
    public constructor() : this(Rules.loadRules())

    /**
     * Convert first name from nominative case to specified case
     *
     * @param firstName  first name at nominative case
     * @param gender gender
     * @param case   case
     * @return converted first name
     */
    public fun firstName(firstName: String, gender: Gender, case: Case): String {
        return convertTo(rules.firstName, firstName, gender, case)
    }

    /**
     * Convert last name from nominative case to specified case
     *
     * @param lastName   last name at nominative case
     * @param gender gender
     * @param case   case
     * @return converted last name
     */
    public fun lastName(lastName: String, gender: Gender, case: Case): String {
        return convertTo(rules.lastName, lastName, gender, case)
    }

    /**
     * Convert middle name from nominative case to specified case
     *
     * @param middleName middle name at nominative case
     * @param gender gender
     * @param case   case
     * @return converted middle name
     */
    public fun middleName(middleName: String, gender: Gender, case: Case): String {
        return convertTo(rules.middleName, middleName, gender, case)
    }

    /**
     * Convert middle name from nominative case to specified case with auto detect gender
     *
     * @param middleName middle name at nominative case
     * @param case   case
     * @return converted middle name
     */
    public fun middleName(middleName: String, case: Case): String {
        val genderCase = Gender.detectGender(middleName)
        return middleName(middleName, genderCase, case)
    }

    private fun convertTo(ruleGroup: RuleGroup, name: String, gender: Gender, case: Case): String {
        val splittedName = name.trim().toLowerCase().split("-").toTypedArray()
        val result = ArrayList<String>(splittedName.size)
        var firstWord = splittedName.size > 1
        for (word in splittedName) {
            val rule: Rule? = ruleGroup.getRule(word, gender, firstWord)
            if (rule == null) {
                result.add(word)
            } else {
                result.add(rule.apply(word, case))
            }
            if (firstWord) {
                firstWord = false
            }
        }
        return result.joinToString("-")
    }
}
